_posts/2023-04-10-opensource-perf.markdown

```diff
@@ -0,0 +1,128 @@
+
+Operating an email server, performance is a concern, as emails are massively used applications, generaly considered critical. Often with even medium sized deployment, handling over 1000 requests per seconds is not uncommon. We use a range of tools for benchmarking the performance of Apache James. Our docs [explains](https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/benchmark/index.adoc) some of the benchmarks we suggest our administrators run to identify performance bottlenecks and malfunctions in their clusters.
+
+Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). Performance testing revealed that the application was spending a noticeable chunk of time running the opensearch-java code, so we set out to examine the performance characteristics of this component.
+
+### Identifying the performance issue
+
+Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that the frequent SPI lookups were causing the performance issue.
+
+[opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. Taken with <https://github.com/jvm-profiling-tools/async-profiler> (From within my applicative container):
+
+`./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1`
+
+
+<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame1.png" alt="Flame graph general overview"/>{: .img-fluid }
+
+
+Then searching for opensearch using the search widget in the top left corner of the flame graph:
+
+<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame2.png" alt="Flame graph: OpenSearch driver threads"/>{: .img-fluid }
+
+-   The driver thread (on the right) with the httpclient event loop
+-   Applicatives calls to the driver from applicative threads (small stacktraces in the center)
```

Comment:
I don't think "applicative" is the right word in English, but I'm not sure what is the right word to use.